[RFC] Fix environment variable on Windows

erw7 opened the pull request:
This PR is another solution for #9681. This may be better because we do not have to maintain `mch_getenv` etc.

However, in `os.getenv` of lua we probably need to make the following settings somewhere:

```lua
os.getenv = function(name)
  return vim.api.nvim_eval('$'..name)
end
```

erw7 committed:
Fix environment variable on Windows

Since uv_os_setenv uses SetEnvironmentVariableW, _wenviron is no
updated. As a result, inconsistency occurs in completion of environment
variable names. Change to use GetEnvironmentStaringsW instead of
_wenviron to solve it.

justinmk commented:
Thank you @erw7 , this is a much nicer solution! 

> However, in os.getenv of lua we probably need to make the following settings somewhere:

That's fine, we can even monkey-patch via the C api if needed (see `nlua_state_init`, for example `nlua_print`).

erw7 committed:
Fix os.getenv of lua on Windows

Change to use os_getenv instead of getenv because environment variable
set by uv_os_setenv can not be get with getenv.

justinmk commented on src/nvim/os/env.c:
```diff
@@ -151,25 +151,39 @@ int os_unsetenv(const char *name)
+  for (wchar_t *it = env; *it != L'\0' || *(it + 1) != L'\0'; it++) {
+    if (index == current_index) {
+      char *utf8_str;
+      int conversion_result = utf16_to_utf8(it, &utf8_str);
```
@erw7 the old code did one call to `utf16_to_utf8()` on the whole `env` string. Is your change a performance optimization?

justinmk commented on src/nvim/lua/executor.c:
```diff
@@ -24,6 +24,10 @@
 #include "nvim/ascii.h"
 
+#ifdef WIN32
+#include "nvim/os/env.h"
```
It's ok to just include `nvim/os/os.h`, we usually treat the `os/` layer as one module.

The new `env.h` file failed `make lint` (specifically the `check-single-includes` test) so I changed this to `nvim/os/os.h`.
